File: plugin-sdk-ts/src/cache.ts
/**
 * Cache management module.
 *
 * Provides access to the cache system for inspection, statistics,
 * and cleanup operations.
 *
 * Requires: `cache_read` and/or `cache_write` permissions.
 */
import { callHostJson } from './host';
import type {
  CacheDetailInfo,
  CacheEntry,
  CacheAccessStats,
  CacheCleanupRecord,
  ExternalCache,
  ExternalCachePath,
  CacheCleanPreview,
  CacheCleanResult,
} from './types';

/** Get detailed cache info. Requires: cache_read */
export function info(): CacheDetailInfo {
  return callHostJson<CacheDetailInfo>('cognia_cache_detail_info', '');
}

/** List cache entries, optionally filtered by type. Requires: cache_read */
export function listEntries(cacheType?: string): CacheEntry[] {
  return callHostJson<CacheEntry[]>(
    'cognia_cache_list_entries',
    JSON.stringify({ cacheType: cacheType ?? null }),
  );
}

/** Get cache access statistics (hit/miss). Requires: cache_read */
export function getAccessStats(): CacheAccessStats {
  return callHostJson<CacheAccessStats>('cognia_cache_get_access_stats', '');
}

/** Get cleanup history records. Requires: cache_read */
export function getCleanupHistory(): CacheCleanupRecord[] {
  return callHostJson<CacheCleanupRecord[]>('cognia_cache_get_cleanup_history', '');
}

/** Discover external caches (system package managers). Requires: cache_read */
export function discoverExternal(): ExternalCache[] {
  return callHostJson<ExternalCache[]>('cognia_cache_discover_external', '');
}

/** Get paths for all known external caches. Requires: cache_read */
export function getExternalPaths(): ExternalCachePath[] {
  return callHostJson<ExternalCachePath[]>('cognia_cache_get_external_paths', '');
}

/** Preview what a cache clean operation would remove. Requires: cache_read */
export function cleanPreview(cacheType: string): CacheCleanPreview {
  return callHostJson<CacheCleanPreview>(
    'cognia_cache_clean_preview',
    JSON.stringify({ cacheType }),
  );
}

/** Clean cache by type. Requires: cache_write */
export function clean(cacheType: string): CacheCleanResult {
  return callHostJson<CacheCleanResult>(
    'cognia_cache_clean',
    JSON.stringify({ cacheType }),
  );
}

File: plugin-sdk/src/cache.rs
//! Cache management module.
//!
//! Provides access to cache inspection, statistics, and cleanup.

use crate::host;
use crate::types::*;
use extism_pdk::*;

/// Get detailed cache info. Requires: cache_read
pub fn info() -> Result<CacheDetailInfo, Error> {
    let result = unsafe { host::cognia_cache_detail_info(String::new())? };
    Ok(serde_json::from_str(&result)?)
}

/// List cache entries. Requires: cache_read
pub fn list_entries(cache_type: Option<&str>) -> Result<Vec<CacheEntry>, Error> {
    let input = serde_json::json!({ "cacheType": cache_type }).to_string();
    let result = unsafe { host::cognia_cache_list_entries(input)? };
    Ok(serde_json::from_str(&result)?)
}

/// Get cache access statistics. Requires: cache_read
pub fn get_access_stats() -> Result<CacheAccessStats, Error> {
    let result = unsafe { host::cognia_cache_get_access_stats(String::new())? };
    Ok(serde_json::from_str(&result)?)
}

/// Get cleanup history records. Requires: cache_read
pub fn get_cleanup_history() -> Result<Vec<CacheCleanupRecord>, Error> {
    let result = unsafe { host::cognia_cache_get_cleanup_history(String::new())? };
    Ok(serde_json::from_str(&result)?)
}

/// Discover external caches. Requires: cache_read
pub fn discover_external() -> Result<Vec<ExternalCache>, Error> {
    let result = unsafe { host::cognia_cache_discover_external(String::new())? };
    Ok(serde_json::from_str(&result)?)
}

/// Get paths for all known external caches. Requires: cache_read
pub fn get_external_paths() -> Result<Vec<ExternalCachePath>, Error> {
    let result = unsafe { host::cognia_cache_get_external_paths(String::new())? };
    Ok(serde_json::from_str(&result)?)
}

/// Preview what a cache clean operation would remove. Requires: cache_read
pub fn clean_preview(cache_type: &str) -> Result<CacheCleanPreview, Error> {
    let input = serde_json::json!({ "cacheType": cache_type }).to_string();
    let result = unsafe { host::cognia_cache_clean_preview(input)? };
    Ok(serde_json::from_str(&result)?)
}

/// Clean cache by type. Requires: cache_write
pub fn clean(cache_type: &str) -> Result<CacheCleanResult, Error> {
    let input = serde_json::json!({ "cacheType": cache_type }).to_string();
    let result = unsafe { host::cognia_cache_clean(input)? };
    Ok(serde_json::from_str(&result)?)
}
